zrml/hybrid-router/src/lib.rs

```diff
@@ -0,0 +1,761 @@
+        ///
+        /// If the maximum price is reached before the entire buy order is filled, the `strategy` parameter
+        /// decides if the order is rolled back (`Strategy::ImmediateOrCancel`) or if a limit order for the
+        /// remaining amount is placed (`Strategy::LimitOrder`).
+        ///
+        /// Complexity: `O(n)`
+        #[pallet::call_index(0)]
+        #[pallet::weight(T::WeightInfo::buy(*asset_count as u32, orders.len() as u32))]
+        #[frame_support::transactional]
+        #[allow(clippy::too_many_arguments)]
+        pub fn buy(
+            origin: OriginFor<T>,
+            market_id: MarketIdOf<T>,
+            #[pallet::compact] asset_count: u16,
+            asset: AssetOf<T>,
+            #[pallet::compact] amount_in: BalanceOf<T>,
+            #[pallet::compact] max_price: BalanceOf<T>,
+            orders: Vec<OrderId>,
+            strategy: Strategy,
+        ) -> DispatchResult {
+            let who = ensure_signed(origin)?;
+
+            Self::do_trade(
+                TxType::Buy,
+                who,
+                market_id,
+                asset_count,
+                asset,
+                amount_in,
+                max_price,
+                &orders,
+                strategy,
+            )?;
+
+            Ok(())
+        }
+
+        /// Routes a sell order to AMM and CDA to achieve the best average execution price.
+        ///
+        /// # Parameters
+        ///
+        /// * `market_id`: The ID of the market to sell on.
+        /// * `asset_count`: The number of assets traded on the market.
+        /// * `asset`: The asset to sell.
+        /// * `amount_in`: The amount of `asset` to sell.
+        /// * `min_price`: The minimum price to sell at.
+        /// * `orders`: A list of orders from the book to use.
+        /// * `strategy`: The strategy to handle the remaining order when the `min_price` is reached.
+        ///
+        /// The elements of `orders` are the orders that the router may use to execute the order. If any of
+        /// these orders are already filled, they are ignored. It is not necessary for the router to use all
+        /// specified orders. The smaller the vector, the larger the risk that the AMM is used to fill large
+        /// chunks of the order.
+        ///
+        /// The `orders` vector **must** be sorted in ascending order by the price of their associated
+        /// orders. Failing this, the behavior of `sell` is undefined.
+        ///
+        /// If the maximum price is reached before the entire buy order is filled, the `strategy` parameter
+        /// decides if the order is rolled back (`Strategy::ImmediateOrCancel`) or if a limit order for the
+        /// remaining amount is placed (`Strategy::LimitOrder`).
+        ///
+        /// Complexity: `O(n)`
+        #[pallet::call_index(1)]
+        #[pallet::weight(T::WeightInfo::sell(*asset_count as u32, orders.len() as u32))]
```

Comment:
I think we should use saturating_into here. This is something that recently popped up in neo-swaps as well. The cast usize to u32 is unsafe on 64 bit systems. This probably has absolutely zero bearing on actual function, but I think using saturating conversion is cleaner.

Reply:
c84cf22739b4a951789fb3f49d5b38db668f52f1